### PR TITLE
Support HCLv2 for Nomad jobs

### PR DESCRIPTION
**Pull request by notnoop**
Support HCLv2 syntax for Nomad jobs.  HCLv2 offers significant expressibility by adding variables and expression supports, and this PR builds the foundation to support that.  In order to experiment, I enabled some functions and dynamic block support to illustrate the power of HCLv2 and to experiment with new syntax in development, but I expect to iterate on these before cutting the final 1.0 release.

## UX and User Visible Changes

This PR updates the CLI and job render API to use HCLv2 parser for Nomad job files.  For the initial draft, I adopted the pPacker Configuration Functions](https://www.packer.io/docs/from-1.5/functions).  Using functions can enable concise HCLv2 jobs - consider the following template blocks:

```hcl
      template {
        destination = "local/motd.txt"
        data        = file("motd.txt")
      }

      template {
        destination = "local/config.json"
        data        = jsonencode({ data = [for s in ["val1", "val2", "val3"] : upper(s)] })
      }
```

Here, Nomad will load the local `motd.txt` and insert it as the template data for the first template.  The second template illustrates the HCLv2 functions, for-looping, as well as function capability: The second template data will be `{"data":["VAL1", "VAL2","VAL3"]}`.

Also, it supports having some input variables to be passed as CLI arguments: `${vars.region}` appearing in job file will be populated by values passed as CLI argument (e.g. `--var region=eu-west-1`).

The exact functions, template, and CLI will iterated on, probably after this PR is merged.

### Backward Compatible Changes

The HCLv2 syntax closely resembles the HCLv1 syntax, so we anticipate that operators can upgrade smoothly. However, there are some notable differences.

In particular, HCLv2 is stricter about block syntax and schema than HCLv1.  HCLv2 requires that block attributes to be unquoted, and doesn't accept map attribute syntax as an alternative to blocks:

```hcl
# valid in HCLv1 and HCLv2
env {
  key1 = "val1"
}

# valid in HCLv1 but NOT HCLv2
env {
  "key1" = "val1"
}

## this is technically a valid HCLv2 but cannot be accepted with above syntax
env = {
  key1 = val1
}
```

To ease adoption, we are making the following concessions:
1. The job run API remain unmodified: frameworks and tools that submit jobs to the job run API endpoint function can function unmodified.  This PR doesn't change the json format the API accept
2. Older CLI submitting jobs to modern Nomad servers with this change also work.  As CLI tools parse and submit the job in the json format, new CLI tools can submit jobs to old servers and old CLI tools can submit jobs to new servers uninterrupted.
3. `nomad job {run|plan|validate}` CLI accept a `-hcl1` flag to force the CLI to use the HCLv1 parser.  We aim to use primarily to ease adoption while operators inspect and modify their job files, and we anticipate to deprecate it soon.

## High-Level Implementation Notes

This PR uses the high-level reflection-based approach for parsing and decoding HCLv2 objects, managed by [`gohcl` package](https://pkg.go.dev/github.com/hashicorp/hcl/v2/gohcl). We use the upstream hclv2 library, but this PR uses some minor tweaks I'll be working to integrating to upstream library.

The PR and implementation can be chunked in three components:

First, tag all `api.Job` structure fields with the appropriate `hcl` tag so they are parsed by HCLv2.  Unlike HCLv1 and mapstructure, `hclv2` gohcl requires that all hcl fields are tagged with name and its hcl syntax type (attribute vs block).  Also, Nomad needs every attribute as optional, as we rely on defaults significantly. With HCLv1, I'm unaware of any case where fields were required that we enforced in the parsing layer.

Second, introduce `jobspec2` package that takes an input stream and decodes into the structs above. Here, we needed to special case some times:
* time.Duration should be converted from strings
* some special types (e.g. api.Constraint, api.Affinity) allow for some alternative equivalent representations like in [constraints distinct_property](https://www.nomadproject.io/docs/job-specification/constraint#distinct_property) .  Custom decoders here override reflection implied schema to perform the needed conversion.
* The Job schema is permissive and allows top level `task` and `vault` even though the api.Job struct doesn't.  We use the custom decoder approach to read these blocks and then canonicalize them as necessary
* task driver Config `map[string]interface{}` : The HCLv2 library stores the entire HCLv2 AST into the map!  We have a custom type that evaluates the map ASTs and decode them into native golang types.

We add small changes to the hcl/v2/gohcl to allow for custom decoders.  The custom decoding hooks are great in our usecase, as we could avoid duplicating api.Job structs as Config specific schemas, code generating hcldec schemas based on api with some custom tweaks, or adding hclv2 code into the API package (which would then leaks hclv2 library as a dependency for api consumers).

Third, we update the CLI and API packages to consume new API.

## Next Steps

After this PR, we plan to iterate on the functions and templating functionality.  This PR supports the same functions as Packer, and only supports HCLv2 native features. I'd advocate for merging this PR with the current functionality, and then getting some operators/users input as well as our own hands-on experience, and then following up with changes.

Secondly, we'll update the guides/docs with HCLv2 details.

## Recommendations for Code Reviewer

First, thank you so much for taking the time to review this PR :).  To ease review, I'd encourage reading the full PR descirption in full and then following through the commits.  The commits roughly map to the components highlighted in the implementation notes section, with some additional auxilary PRs.

Fixes https://github.com/hashicorp/nomad/issues/7220

### CODE REVIEW

**Comment by tgross** [comment]
@notnoop I'm still chewing on this... but just for my understanding: this PR has `hcl` struct tags on fields that we don't (under HCL1) have as part of the jobspec but are returned in the API. (For example [`Job.ParentID`](https://github.com/hashicorp/nomad/pull/9142/files#diff-af7c576cce844bb1b9ab602a80eb688fefeb12d84075fd7f54a6cfb39f9c4620R796)) Are these just required for serialization or does adding tags to these fields let them get parsed in from jobspecs?

**Comment by notnoop** [comment]
> this PR has `hcl` struct tags on fields that we don't (under HCL1) have as part of the jobspec but are returned in the API. (For example [`Job.ParentID`](https://github.com/hashicorp/nomad/pull/9142/files#diff-af7c576cce844bb1b9ab602a80eb688fefeb12d84075fd7f54a6cfb39f9c4620R796)) Are these just required for serialization or does adding tags to these fields let them get parsed in from jobspecs?

Interesting.  We probably shouldn't be tagging these fields, if we don't expect them in the job hcl body.  Looks like HCLv1 parses them as HCLv1 parser doesn't check unexpected fields, so we might silently populating these field but then dropping them.  I guess that's an argument for wanting to have an explicit JobConfig type that doesn't contain any fields populated by server.

File: api/jobs.go [hunk 1]
```diff
@@ -779,38 +782,43 @@ func (p *PeriodicConfig) GetLocation() (*time.Location, error) {
...
-	Periodic          *PeriodicConfig
-	ParameterizedJob  *ParameterizedJobConfig
 	Dispatched        bool
 	Payload           []byte
-	Reschedule        *ReschedulePolicy
-	Migrate           *MigrateStrategy
-	Meta              map[string]string
-	ConsulToken       *string `mapstructure:"consul_token"`
-	VaultToken        *string `mapstructure:"vault_token"`
 	VaultNamespace    *string `mapstructure:"vault_namespace"`
```
Comment (notnoop) [inline api/jobs.go:822]: May require follow up for VaultNamespace. I'm a little surprised that vault_namespace field isn't a valid config option. I checked the field is rejected emperically, and confirmed it's not listed as a valid key in the hclv1 parser: https://github.com/hashicorp/nomad/blob/1d6997942992e92a0eccaec1d87cb0c1d43eeb0b/jobspec/parse_job.go#L79-L82.

File: jobspec2/hclutil/blockattrs.go [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+// More concretely, it changes the following:
+//
+// ```
+// config {
+//   meta { ... }
+// }
+// ```
+// to
+//
+// ```
```
Comment (tgross) [inline jobspec2/hclutil/blockattrs.go:23]: Maybe a copy-paste error in this example?

File: jobspec2/hclutil/blockattrs.go [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+func (b *blockAttrs) JustAttributes() (hcl.Attributes, hcl.Diagnostics) {
+	body, ok := b.body.(*hcls.Body)
+	if !ok {
+		return b.body.JustAttributes()
+	}
+
+	attrs := make(hcl.Attributes)
+	var diags hcl.Diagnostics
+
+	if body.Attributes == nil && len(body.Blocks) == 0 {
```
Comment (tgross) [inline jobspec2/hclutil/blockattrs.go:81]: Does returning early get us anything here? Both will `x == nil` and `len(x) == 0` if empty, in which case there's nothing to range over.

File: jobspec2/parse_job.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+	if j.Periodic != nil && j.Periodic.Spec != nil {
+		v := "cron"
+		j.Periodic.SpecType = &v
+	}
+
+	normalizeVault(jw.Extra.Vault)
+
+	if len(jw.Extra.Tasks) != 0 {
+		alone := make([]*api.TaskGroup, 0, len(jw.Extra.Tasks))
```
Comment (tgross) [inline jobspec2/parse_job.go:65]: Huh, TIL we allowed tasks without a task group!

File: jobspec2/parse_test.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+	"io/ioutil"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/hashicorp/nomad/jobspec"
+	"github.com/stretchr/testify/require"
+)
+
+func TestEquivalentToHCL1(t *testing.T) {
```
Comment (tgross) [inline jobspec2/parse_test.go:13]: As a bonus, this is a nice way of saving us from writing a whole lot of test code!

File: jobspec2/hcl_conversions.go [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+
+var hclDecoder *gohcl.Decoder
+
+func init() {
+	hclDecoder = newHCLDecoder()
+	hclDecoder.RegisterBlockDecoder(reflect.TypeOf(api.TaskGroup{}), decodeTaskGroup)
+}
+
+func newHCLDecoder() *gohcl.Decoder {
+	decoder := &gohcl.Decoder{}
```
Comment (tgross) [inline jobspec2/hcl_conversions.go:24]: Obviously this compiles but just for my education where is this `Decoder` struct coming from? It's not in the [docs](https://pkg.go.dev/github.com/hashicorp/hcl/v2@v2.7.0/gohcl) and I can't find it in the `gohcl` source either!

File: command/helpers.go [hunk 7]
```diff
@@ -378,20 +380,28 @@ READ:
...
 	var jobfile io.Reader
+	pathName := filepath.Base(jpath)
 	switch jpath {
 	case "-":
 		if j.testStdin != nil {
 			jobfile = j.testStdin
 		} else {
 			jobfile = os.Stdin
 		}
+		pathName = "stdin.hcl"
```
[the remark this answers is not in the document]
Comment (notnoop) [reply command/helpers.go:404]: I'll use `-` here, as it's will be more obvious to the user.  As we don't rely on the filename, we should be fine, and the filename is only used for reporting errors to users.

File: jobspec2/hclutil/blockattrs.go (hunk 3 above)
Comment (notnoop) [reply jobspec2/hclutil/blockattrs.go:81]: removing it - meant it to be fast-path optimization but seeing it now, it's not particularly useful.

File: jobspec2/hcl_conversions.go (hunk 6 above)
Comment (notnoop) [reply jobspec2/hcl_conversions.go:24]: Yeah, I'm using a slightly modified hcl/v2 sha now - and will be working to getting the changes incorporated in hcl/v2 branch.  You can review those changes in https://github.com/hashicorp/hcl/compare/hcl2...hcl2-nomad-teaks .

**Pull request merged**